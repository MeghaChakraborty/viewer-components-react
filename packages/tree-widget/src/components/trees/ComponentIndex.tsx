/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import * as React from "react";
import { Ruleset } from "@bentley/presentation-common";
import { ControlledTreeWrapper, populateMapWithCommonMenuItems } from "./TreeWithRuleset";
import { FunctionIconInfo, TreeNodeFunctionIconInfoMapper } from "./FunctionalityProviders";
import { IPresentationTreeDataProvider, PresentationTreeDataProvider } from "@bentley/presentation-components";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import componentIndex from "../rulesets/ComponentIndex.json";
import { TreeWidget as BuildingUIComponents } from "../../TreeWidget";
import { GenericOptionItemHandler, OptionItemHandler } from "./OptionItemHandlers";
import { LoadableRuleSetComponent } from "./LoadableRuleSetComponent";

export interface ComponentIndexProps {
  iModel: IModelConnection;
  displayGuids: boolean;
  setIsDisplayGuids: (displayGuids: boolean) => void;
  enableVisibility?: boolean;
}
const COMPONENT_INDEX_NAME = "ComponentIndex";

export const ComponentIndex: React.FC<ComponentIndexProps> = (props: ComponentIndexProps) => {
  const treeName = "ComponentIndexTree";
  const dataProvider = React.useMemo(() => {
    const provider = new PresentationTreeDataProvider({
      imodel: props.iModel,
      ruleset: componentIndex.id,
    });
    provider.pagingSize = 20; // paging size is now needed for the controlled tree.
    return provider;
  }, [props.iModel],
  );

  const { functionIconMapper, optionItems, displayGuidHandler } = React.useMemo(() => {
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const optionItems: OptionItemHandler[] = [];

    populateContextMenuItems(treeName, functionIconMapper, dataProvider);

    const displayGuidHandler = new GenericOptionItemHandler("Show Guids", BuildingUIComponents.translate("contextMenu.showGuids"), "icon-label", () => { return props.displayGuids; }, props.setIsDisplayGuids);
    optionItems.push(displayGuidHandler);
    return { functionIconMapper, optionItems, displayGuidHandler };
  }, [dataProvider]);

  displayGuidHandler._getItemState = () => props.displayGuids;

  const componentIndexTree = React.useMemo(() => (<ControlledTreeWrapper
    iModel={props.iModel}
    loadedRuleset={componentIndex as Ruleset}
    dataProvider={dataProvider}
    treeName={COMPONENT_INDEX_NAME} treeNodeIconMapper={functionIconMapper} optionItems={optionItems}
    searchTools={false}
    displayGuids={props.displayGuids}
    setIsDisplayGuids={props.setIsDisplayGuids}
    enableVisibility={props.enableVisibility ? props.enableVisibility : false}
  />), [props.iModel, props.displayGuids, props.setIsDisplayGuids]);

  return (
    <LoadableRuleSetComponent ruleSet={componentIndex as Ruleset} >
      {componentIndexTree}
    </LoadableRuleSetComponent>
  );
}

function populateContextMenuItems(treeName: string, mapper: TreeNodeFunctionIconInfoMapper, dataProvider: IPresentationTreeDataProvider) {
  populateMapWithCommonMenuItems(treeName, mapper, dataProvider, componentIndex.id);
}
