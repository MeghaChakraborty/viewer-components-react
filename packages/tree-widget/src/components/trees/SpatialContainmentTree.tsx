/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { Presentation } from "@bentley/presentation-frontend";
import spatialRulesDefault from "../rulesets/SpatialBreakdown.json";
import spatialRulesByType from "../rulesets/SpatialBreakdownByType.json";
import { IModelApp, IModelConnection, SpatialViewState } from "@bentley/imodeljs-frontend";
import { IPresentationTreeDataProvider, PresentationTreeDataProvider } from "@bentley/presentation-components";
import { Ruleset } from "@bentley/presentation-common";
import { populateMapWithCommonMenuItems, ControlledTreeWrapper } from "./TreeWithRuleset";
import { TreeWidget as BuildingUIComponents } from "../../TreeWidget";
import { BuildingClipPlanesProvider, ClearSectionsFunctionalityProvider, CombinedTreeNodeFunctionalityProvider, SpaceClipPlanesProvider, StoryClipPlanesProvider, TreeNodeFunctionIconInfoMapper, FunctionIconInfo } from "./FunctionalityProviders";
import { ClipAtSpacesHandler, GenericOptionItemHandler, LabelHandler, TopViewHandler, OptionItemHandler } from "./OptionItemHandlers";
import { LoadableRuleSetComponent } from "./LoadableRuleSetComponent";

export interface SpatialContainmentTreeProps {
  iModel: IModelConnection;
  displayGuids: boolean;
  setIsDisplayGuids: (displayGuids: boolean) => void;
  enableVisibility?: boolean;
  clipHeight?: number;
  clipAtSpaces?: boolean;
  displayByType: boolean;
  setDisplayByType: (displayByType: boolean) => void;
}

function updateModelsInRulesetVars() {
  let viewState = IModelApp.viewManager.selectedView ?.view as SpatialViewState;
  let modelIds = Array.from(viewState.modelSelector.models);
  Presentation.presentation.vars("ui-framework/SpatialBreakdown").setId64s("displayed_model_ids", modelIds);
}

const _onModelsChanged = async () => {
  updateModelsInRulesetVars();
}

// <ConnectedSimpleTreeWithRuleset ruleSet={spatialRules as Ruleset}  controller={treeWithRulesetController} />
export const SpatialContainmentTree: React.FC<SpatialContainmentTreeProps> = (props: SpatialContainmentTreeProps) => {
  const treeName = "SpatialContainmentTree";
  let spatialRules: any;
  if (props.displayByType) {
    spatialRules = spatialRulesByType;
  } else {
    spatialRules = spatialRulesDefault;
  }

  React.useEffect(() => {
    if (IModelApp.viewManager.selectedView ?.view.isSpatialView) {
      IModelApp.viewManager.selectedView.onViewedModelsChanged.addListener(_onModelsChanged);
      updateModelsInRulesetVars();
    }
    return () => {
      if (IModelApp.viewManager.selectedView ?.view.isSpatialView) {
        IModelApp.viewManager.selectedView.onViewedModelsChanged.removeListener(_onModelsChanged);
      }
    }
  });

  const dataProvider = React.useMemo(() => {
    const dataProviderInner = new PresentationTreeDataProvider({ imodel: props.iModel, ruleset: spatialRules.id, appendChildrenCountForGroupingNodes: true });
    dataProviderInner.pagingSize = 20; // paging size is now needed for the controlled tree.
    return dataProviderInner;
  }, [props.iModel, props.displayByType]);

  const { functionIconMapper, optionItems, displayGuidHandler, displayByTypeHandler } = React.useMemo(() => {
    const functionIconMapper = new TreeNodeFunctionIconInfoMapper(dataProvider);
    const optionItems: OptionItemHandler[] = [];

    populateContextAndOptionMenuItems(treeName, functionIconMapper, optionItems, dataProvider, spatialRules, props.clipHeight, props.clipAtSpaces);

    const displayGuidHandler = new GenericOptionItemHandler("Show Guids", BuildingUIComponents.translate("contextMenu.showGuids"), "icon-label", () => props.displayGuids, props.setIsDisplayGuids);
    optionItems.push(displayGuidHandler);
    const displayByTypeHandler = new GenericOptionItemHandler("Display By Type", BuildingUIComponents.translate("contextMenu.displayByType"), "icon-hierarchy-tree", () => props.displayByType, props.setDisplayByType);
    optionItems.push(displayByTypeHandler);

    return { functionIconMapper, optionItems, displayGuidHandler, displayByTypeHandler };
  }, [dataProvider]); // the GenericOptionItemHandler uses function to retrieve state - don't need to update

  displayGuidHandler._getItemState = () => props.displayGuids;
  displayByTypeHandler._getItemState = () => props.displayByType;
  const containmentTree = React.useMemo(() => <ControlledTreeWrapper iModel={props.iModel} loadedRuleset={spatialRules as Ruleset} dataProvider={dataProvider}
    treeName={treeName} treeNodeIconMapper={functionIconMapper} optionItems={optionItems} searchTools={true}
    displayGuids={props.displayGuids} setIsDisplayGuids={props.setIsDisplayGuids} enableVisibility={props.enableVisibility ? props.enableVisibility : false} />, [props.iModel, props.displayGuids, props.setIsDisplayGuids, functionIconMapper, props.displayByType]);

  return (<LoadableRuleSetComponent ruleSet={spatialRules as Ruleset}>
    {containmentTree}
  </LoadableRuleSetComponent>);
}

function populateContextAndOptionMenuItems(treeName: string, mapper: TreeNodeFunctionIconInfoMapper, optionItems: OptionItemHandler[], dataProvider: IPresentationTreeDataProvider, spatialRules: any, clipHeight?: number, clipAtSpaces?: boolean) {
  populateMapWithCommonMenuItems(treeName, mapper, dataProvider, spatialRules.id);

  const combinedFunctionalityProvider = new CombinedTreeNodeFunctionalityProvider(treeName, dataProvider);
  const storyClipSectionProvider = new StoryClipPlanesProvider(treeName, dataProvider, false, false, clipHeight);
  const spaceClipSectionProvider = new SpaceClipPlanesProvider(treeName, dataProvider, false, clipHeight);
  const buildingClipSectionProvider = new BuildingClipPlanesProvider(treeName, dataProvider, false);
  combinedFunctionalityProvider.setFunctionalityProviderForClass("BuildingSpatial:Story", storyClipSectionProvider);
  combinedFunctionalityProvider.setFunctionalityProviderForClass("BuildingSpatial:Space", spaceClipSectionProvider);
  combinedFunctionalityProvider.setFunctionalityProviderForClass("BuildingSpatial:Building", buildingClipSectionProvider);

  mapper.registerForNodesOfClasses(["BuildingSpatial:Story", "BuildingSpatial:Space", "BuildingSpatial:Building"], {
    key: "Create section planes",
    label: BuildingUIComponents.translate("contextMenu.createSectionPlanes"),
    functionalityProvider: combinedFunctionalityProvider,
    toolbarIcon: "icon-section-tool",
  });
  mapper.registerGlobal({
    key: "Clear section planes",
    label: BuildingUIComponents.translate("contextMenu.clearSectionPlanes"),
    functionalityProvider: new ClearSectionsFunctionalityProvider(treeName, dataProvider),
    toolbarIcon: "icon-section-clear",
  });

  const labelHandler = new LabelHandler(storyClipSectionProvider, "Toggle Space Labels", BuildingUIComponents.translate("contextMenu.toggleSpaceLabels"), "icon-text");
  optionItems.push(labelHandler);

  const topViewHandler = new TopViewHandler([storyClipSectionProvider, spaceClipSectionProvider, buildingClipSectionProvider], "Top Down View", BuildingUIComponents.translate("contextMenu.topDownView"), "icon-cube-faces-top");
  optionItems.push(topViewHandler);

  if (clipAtSpaces !== undefined) {
    const clipAtSpacesHandler = new ClipAtSpacesHandler(storyClipSectionProvider, clipAtSpaces, "Clip At Spaces", BuildingUIComponents.translate("contextMenu.clipAtSpaces"), "icon-section-tool");
    optionItems.push(clipAtSpacesHandler);
  }
}
