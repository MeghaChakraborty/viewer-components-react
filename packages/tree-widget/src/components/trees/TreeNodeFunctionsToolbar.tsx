/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { IconButton } from "./IconButton";
import * as React from "react";
import { TreeModel, TreeModelNode } from "@bentley/ui-components";
import { TreeNodeFunctionIconInfoMapper, FunctionIconInfo } from "./FunctionalityProviders";


export interface TreeNodeFunctionsToolbarProps {
  treeModel: TreeModel;
  treeNodeIconMapper: TreeNodeFunctionIconInfoMapper;
  selectedNodes: TreeModelNode[];
}

export const TreeNodeFunctionsToolbar: React.FC<TreeNodeFunctionsToolbarProps> = (props: TreeNodeFunctionsToolbarProps) => {
  const [toolBarIcons, setToolbarIcons] = React.useState<React.ReactNode[] | undefined>(undefined);

  React.useEffect( () => {
    getToolbarIcons(props.treeNodeIconMapper, props.selectedNodes, props.treeModel)
      .then((icons) => setToolbarIcons(icons))
  } , [props.selectedNodes]);

  return (<div className="custom-tree-toolbar">
            {toolBarIcons}
          </div>);
};


async function getToolbarIcons(iconMapper: TreeNodeFunctionIconInfoMapper, selectedNodes: TreeModelNode[], treeModel: TreeModel ) {
  const items: React.ReactNode[] = [];
  let functionIconInfos: ReadonlyArray<FunctionIconInfo>;
  if (selectedNodes.length === 1){
    functionIconInfos = await iconMapper.getFunctionIconInfosFor(selectedNodes[0]);
  } else {
    // treat multiple selection same as none selected for now
    functionIconInfos = await iconMapper.getFunctionIconInfosFor();
  }
  functionIconInfos.forEach((info: FunctionIconInfo) => {
    items.push(
      <IconButton
              key= {info.key}
              icon= {info.toolbarIcon}
              className={"toolbar-icon"}
              disabled={info.disabled !== false}
              onClick={() => { info.functionalityProvider.performAction(selectedNodes[0], treeModel); }}
              title={info.label}
            />,
    );
  });
  return items;
}
