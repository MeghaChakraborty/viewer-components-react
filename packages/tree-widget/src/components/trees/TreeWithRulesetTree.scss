/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

@import "~@bentley/ui-core/lib/ui-core/index";
@mixin ui-fw-visibility-tree-base {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;
  position: relative;

  .core-tree-node.with-checkbox {
    > .contents {
      padding-left: 31px;
    }
  }

  .core-image-checkbox {
    position: absolute;
    left: 5px;
    padding: 0;
    margin: -2px 0px 0px 5px;
  }

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

  .components-tree-errormessage {
    position: absolute;
    top: 40%;
    left: 50%;
    width: 75%;
    text-align: center;
    transform: translate(-50%,-50%);
    font-style: italic;
    color: $buic-foreground-muted;

    .errormessage-header {
      display: block;
      font-weight: bold;
    }
  }

}

.dr-spatial-tree {
  display: flex;
  flex-direction: column;
  height: 100%;
}
.components-tree-node-content {
  height: $uicore-l;
  line-height: $uicore-font-size-small;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.custom-tree-content {
  height: 100%;
  display: flex;
  flex-direction: column;
  position: relative;
}

.custom-tree-container {
  @include ui-fw-visibility-tree-base;

  height: 100%;

  .custom-tree-node {
    display: flex;
    align-items: center;

    &:hover {
      background: $buic-row-hover;
    }

    &.selected {
      background: $buic-row-selection;
    }

    .custom-tree-node-label {
      display: flex;
      align-items: center;
      text-overflow: ellipsis;
      white-space: nowrap;
      flex-wrap: nowrap;
      flex: 1;

      > .custom-tree-node-label-wrapper {
        height: 24px;
        display: flex;
        align-items: center;
      }
    }
  }
}

.tree-item-wrapper{
  display: flex;
  flex: $uicore-xxs;
  flex-direction: row;
  .tree-item-left-border {
    display: flex;
    width: $uicore-xs;
    position: sticky;
    left: 0px; top: 0px;
    height: inherit;
  }
  .tree-item {
    display: flex;
    flex: 1$uicore-xxs;
    flex-direction: column;
  }
}


.custom-tree-toolbar {
  display: flex;
  align-items: center;
  .right {
    justify-content: flex-end;
  }
}

.toolbar-icon {
  color: $buic-icon-color;
}
.meatballButtonContainer {
  margin-left: auto;
}
.icon-checkmark {
  margin-top: 5px;
  margin-right: 6px;
}

.custom-tree-popup {
  border: 1px solid $buic-inputs-border;
  padding: $uicore-s;

  .custom-tree-popup-container {
    display: flex;
    align-items: center;
  }
}

.popup {
  width: 195px;
  display: flex;
  overflow: hidden;
}
.item-container {
  width: 195px;
  height: 25px;
  display: inline-flex;
}
.Popup-item {
  display: inline-flex;
  align-items: center;
  vertical-align: middle;
  width: 195px;
  height: 25px;
  justify-content: left;
  cursor: pointer;
  font-size: 14px;

  * {
    margin-left: 5px;
  }
}
