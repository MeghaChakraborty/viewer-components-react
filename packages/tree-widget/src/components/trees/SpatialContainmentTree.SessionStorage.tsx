/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { SpatialContainmentTree } from "./SpatialContainmentTree";
import { IModelConnection } from "@bentley/imodeljs-frontend";

export interface SessionStorageSpatialContainmentTreeProps {
  iModel: IModelConnection;
}

const SHOW_GUID_STORAGE_KEY = "discard-guid-from-label";
const DISPLAY_BY_TYPE_STORAGE_KEY = "display-by-type";

export const SessionStorageSpatialContainmentTree: React.FC<SessionStorageSpatialContainmentTreeProps> = (props: SessionStorageSpatialContainmentTreeProps) => {
  const [displayGuids, setDisplayGuids] = React.useState<boolean>(sessionStorage.getItem(SHOW_GUID_STORAGE_KEY) !== "false");
  const [displayByType, setDisplayByType] = React.useState<boolean>(sessionStorage.getItem(DISPLAY_BY_TYPE_STORAGE_KEY) === "true");

  const setIsDisplayByType = React.useCallback((value: boolean) => {
    sessionStorage.setItem(DISPLAY_BY_TYPE_STORAGE_KEY, String(value));
    setDisplayByType(value);
  }, []);

  const setIsDisplayGuids = React.useCallback((value: boolean) => {
    sessionStorage.setItem(SHOW_GUID_STORAGE_KEY, String(value))
    setDisplayGuids(value);
  }, []);

  return <SpatialContainmentTree iModel={props.iModel} displayByType={displayByType} setDisplayByType={setIsDisplayByType} displayGuids={displayGuids} setIsDisplayGuids={setIsDisplayGuids} enableVisibility={false} />;
}
