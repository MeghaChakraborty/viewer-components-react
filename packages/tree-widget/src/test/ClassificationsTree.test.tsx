/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/



import * as React from "react";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import { IModelApp, NoRenderApp, IModelConnection } from "@bentley/imodeljs-frontend";
import { BeEvent } from "@bentley/bentleyjs-core";
import { TestUtils, TreeWithRuleSetController, setupDataProvider } from "./Utils";
import { Presentation } from "@bentley/presentation-frontend";
import * as moq from "typemoq";
import { PresentationManager, RulesetManager, RulesetVariablesManager } from "@bentley/presentation-frontend";
import { ClassificationsTree } from "../components/trees/ClassificationsTree";

describe("TreeWithRuleset Component tests.", () => {

  const connection = moq.Mock.ofType<IModelConnection>();
  const controller = moq.Mock.ofType<TreeWithRuleSetController>();
  const shutdownIModelApp = () => {
    if (IModelApp.initialized)
      IModelApp.shutdown();
  };
  before(async () => {
    shutdownIModelApp();
    NoRenderApp.startup();
    Presentation.terminate();
    Presentation.initialize();
    await TestUtils.initializeUiFramework(connection.object);
    IModelApp.i18n.registerNamespace("TreeWidget");
    controller.setup((x) => x.createDataProvider()).returns(() => setupDataProvider(connection.object));
  });


  const rulesetManagerMock = moq.Mock.ofType<RulesetManager>();
  const rulesetVariablesManagerMock = moq.Mock.ofType<RulesetVariablesManager>();
  const presentationManagerMock = moq.Mock.ofType<PresentationManager>();

  beforeEach(async () => {
    rulesetVariablesManagerMock.reset();
    presentationManagerMock.reset();
    presentationManagerMock.setup((x) => x.rulesets()).returns(() => rulesetManagerMock.object);
    presentationManagerMock.setup((x) => x.vars(moq.It.isAny())).returns(() => rulesetVariablesManagerMock.object);
    rulesetVariablesManagerMock.setup((x) => x.onVariableChanged).returns(() => new BeEvent());
    Presentation.setPresentationManager(presentationManagerMock.object);
    await Presentation.initialize();
  });

  afterEach(() => {
    Presentation.terminate();
  });

  after(() => {
    TestUtils.terminateUiFramework();
    Presentation.terminate();
    shutdownIModelApp();
  });

  it("ClassificationsTree renders correctly with children", () => {
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ClassificationsTree iModel={connection.object} displayGuids={true} setIsDisplayGuids={(_display: boolean) => { }} enableVisibility={false} />
      </Provider>
    );
    mount(wrapper).should.matchSnapshot();
  });

  it("ClassificationsTree renders correctly with visibility", () => {
    const wrapper = (
      <Provider store={TestUtils.store}>
        <ClassificationsTree iModel={connection.object} displayGuids={true} setIsDisplayGuids={(_display: boolean) => { }} enableVisibility={true} />
      </Provider>
    );
    mount(wrapper).should.matchSnapshot();
  });
});
