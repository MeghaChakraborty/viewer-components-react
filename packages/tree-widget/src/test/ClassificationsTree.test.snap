// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeWithRuleset Component tests. ClassificationsTree renders correctly with children 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    displayGuids={true}
    enableVisibility={false}
    iModel={null}
    setIsDisplayGuids={[Function]}
  >
    <Component
      ruleSet={
        Object {
          "id": "ui-framework/ClassificationSystems",
          "rules": Array [
            Object {
              "ruleType": "RootNodes",
              "specifications": Array [
                Object {
                  "arePolymorphic": true,
                  "classes": Array [
                    Object {
                      "classNames": Array [
                        "ClassificationSystem",
                      ],
                      "schemaName": "ClassificationSystems",
                    },
                  ],
                  "groupByClass": false,
                  "groupByLabel": false,
                  "specType": "InstanceNodesOfSpecificClasses",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"ClassificationSystem\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "doNotSort": true,
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "ClassificationTable",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ClassificationSystemOwnsClassificationTable",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"ClassificationTable\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "hideNodesInHierarchy": true,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Model",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ModelModelsElement",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Model\\", \\"BisCore\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Classification",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ModelContainsElements",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Classification\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Classification",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ClassificationOwnsSubClassifications",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"TypeDefinitionElement\\", \\"BisCore\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "PhysicalElement",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "GeometricElement3dHasTypeDefinition",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Classification\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Element",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ElementHasClassifications",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
              "ruleType": "Content",
              "specifications": Array [
                Object {
                  "relationshipPaths": Array [
                    Array [
                      Object {
                        "count": "*",
                        "direction": "Forward",
                        "relationship": Object {
                          "className": "ClassificationSystemOwnsClassificationTable",
                          "schemaName": "ClassificationSystems",
                        },
                      },
                    ],
                  ],
                  "specType": "ContentRelatedInstances",
                },
              ],
            },
          ],
        }
      }
    >
      <div />
    </Component>
  </Component>
</Provider>
`;

exports[`TreeWithRuleset Component tests. ClassificationsTree renders correctly with visibility 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Component
    displayGuids={true}
    enableVisibility={true}
    iModel={null}
    setIsDisplayGuids={[Function]}
  >
    <Component
      ruleSet={
        Object {
          "id": "ui-framework/ClassificationSystems",
          "rules": Array [
            Object {
              "ruleType": "RootNodes",
              "specifications": Array [
                Object {
                  "arePolymorphic": true,
                  "classes": Array [
                    Object {
                      "classNames": Array [
                        "ClassificationSystem",
                      ],
                      "schemaName": "ClassificationSystems",
                    },
                  ],
                  "groupByClass": false,
                  "groupByLabel": false,
                  "specType": "InstanceNodesOfSpecificClasses",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"ClassificationSystem\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "doNotSort": true,
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "ClassificationTable",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ClassificationSystemOwnsClassificationTable",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"ClassificationTable\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "hideNodesInHierarchy": true,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Model",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ModelModelsElement",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Model\\", \\"BisCore\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Classification",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ModelContainsElements",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Classification\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Classification",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ClassificationOwnsSubClassifications",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Forward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"TypeDefinitionElement\\", \\"BisCore\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "PhysicalElement",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "GeometricElement3dHasTypeDefinition",
                    ],
                    "schemaName": "BisCore",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ParentNode.IsOfClass(\\"Classification\\", \\"ClassificationSystems\\")",
              "ruleType": "ChildNodes",
              "specifications": Array [
                Object {
                  "groupByClass": false,
                  "groupByLabel": false,
                  "relatedClasses": Object {
                    "classNames": Array [
                      "Element",
                    ],
                    "schemaName": "BisCore",
                  },
                  "relationships": Object {
                    "classNames": Array [
                      "ElementHasClassifications",
                    ],
                    "schemaName": "ClassificationSystems",
                  },
                  "requiredDirection": "Backward",
                  "specType": "RelatedInstanceNodes",
                },
              ],
            },
            Object {
              "condition": "ContentDisplayType = \\"RelatedElementsRequest\\"",
              "ruleType": "Content",
              "specifications": Array [
                Object {
                  "relationshipPaths": Array [
                    Array [
                      Object {
                        "count": "*",
                        "direction": "Forward",
                        "relationship": Object {
                          "className": "ClassificationSystemOwnsClassificationTable",
                          "schemaName": "ClassificationSystems",
                        },
                      },
                    ],
                  ],
                  "specType": "ContentRelatedInstances",
                },
              ],
            },
          ],
        }
      }
    >
      <div />
    </Component>
  </Component>
</Provider>
`;
